Tests/ConfidenceProviderTests/ConfidenceFeatureProviderTest.swift

```diff
@@ -26,38 +26,69 @@ class ConfidenceFeatureProviderTest: XCTestCase {
     }
 
     func testSlowFirstResolveWillbeCancelledOnSecondResolve() async throws {
-        class FakeClient: ConfidenceResolveClient {
+        let expectation1 = expectation(description: "First resolve completed")
+        let expectation2 = expectation(description: "Second resolve should never complete")
+        expectation2.isInverted = true
+        let expectation3 = expectation(description: "Third resolve completed")
+        let expectation4 = expectation(description: "Unlock second resolve")
+
+        class FakeClient: XCTestCase, ConfidenceResolveClient {
             var callCount = 0
             var returnCount = 0
+            let expectation1: XCTestExpectation
+            let expectation2: XCTestExpectation
+            let expectation3: XCTestExpectation
+            let expectation4: XCTestExpectation
+
+            init(expectation1: XCTestExpectation,
+                 expectation2: XCTestExpectation,
+                 expectation3: XCTestExpectation,
+                 expectation4: XCTestExpectation) {
+                self.expectation1 = expectation1
+                self.expectation2 = expectation2
+                self.expectation3 = expectation3
+                self.expectation4 = expectation4
+                super.init(invocation: nil) // Workaround to use expectations in test
+            }
 
             func resolve(ctx: ConfidenceStruct) async throws -> ResolvesResult {
                 callCount += 1
-                if callCount == 1 {
-                    // wait 2 seconds
-                    try await Task.sleep(nanoseconds: UInt64(1 * Double(NSEC_PER_MSEC)))
-                    returnCount += 1
-                } else {
-                    returnCount += 1
+                switch callCount {
+                case 1:
+                    expectation1.fulfill()
+                case 2:
+                    print(Task.isCancelled)
+                    await fulfillment(of: [expectation4], timeout: 5.0)
+                    print(Task.isCancelled) // It is true but code still runs!
+                    expectation2.fulfill() // Expectation is inverted, should never arrive here
```

Comment:
I feel more comfortable about racing issue if we throw and fail the test here

Reply:
Yep, much nicer with XCTFail 👍